Add a section about running your first real experiment. (#33)
beaker/docs: docs/start/experiment.md

# Your First Real Beaker Experiment

By now you are probably eager to run a real experiment in Beaker.  This
tutorial will take you through the steps required to run your experiment.

> This tutorial assumes that your experiment will need GPUs, built off of
> PyTorch, and consists of a single command.  If these assumptions don't hold for
> your experiments, then these general steps will still apply but you might need
> to make some more modifications to actually run your experiment.

Reviewer: ```suggestion
This tutorial assumes that your experiment needs GPUs, is built with
PyTorch, and consists of a single command.  If these assumptions don't hold for
your experiments, then these general steps will still apply but you might need
to make some more modifications to actually run your experiment.
```

> This example assumes you've successfully completed [Beaker and Docker installation](install.md).

Reviewer: ```suggestion
This example assumes you've successfully [installed Beaker and Docker](install.md).
```

> ## Creating a Docker image for your code

Reviewer: ```suggestion
## Creating a Docker Image for Your Code
```

As a first step, you need to create a Docker image for your code so Beaker can run it on a cluster.
Fortunately, for simple Python applications this is quite straightforward.  The first step is to
create a `Dockerfile` in the same directory as your code.  Below is a sample Dockerfile that can

> be adapted to your project.  It was simplified from https://github.com/allenai/allennlp/blob/main/Dockerfile.

Reviewer: ```suggestion
be adapted to your project.  It was simplified from [this one](https://github.com/allenai/allennlp/blob/main/Dockerfile).
```

```
# The base image, which will be the starting point for the Docker image.

> # The python:3.8 is Ubuntu with Python installed.

Reviewer: ```suggestion
# This image includes Ubuntu with the latest Python 3.8 release installed.
```

FROM python:3.8

ENV LC_ALL=C.UTF-8
ENV LANG=C.UTF-8

# This section is needed for GPUs to work.
ENV LD_LIBRARY_PATH /usr/local/nvidia/lib:/usr/local/nvidia/lib64

ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES compute,utility
LABEL com.nvidia.volumes.needed="nvidia_driver"

# This is the directory that files will be copied into.
# It's also the directory that you'll start in if you connect to the image.
WORKDIR /stage/

> # Copy the `requirements.txt` to `/stage/requirements.txt/` and then install them.

Reviewer: ```suggestion
# Copy the `requirements.txt` to `/stage/requirements.txt` and then install them.
```

# We do this first because it's slow and each of these commands are cached in sequence.
COPY requirements.txt .
RUN pip install -r requirements.txt

# Copy the file `main.py` to `/stage/main.py`
# You might need multiple of these statements to copy all the files you need for your experiment.
COPY main.py .

# Copy the folder `scripts` to `scripts/`
# You might need multiple of these statements to copy all the folders you need for your experiment.

> COPY scripts scripts/

Reviewer: I wonder if we should consider using `COPY . .` with a `.dockerignore` file?

Maybe that's more complicated than a series of separate `COPY` statements? On one hand I like the explicitness of the way it is now, but it's also a little tedious.

In fact, I wonder if we could provide a template that just has a `.dockerignore` with the relevant things in place `*.pyc` files n' such).

Author: I do this myself and I find it convenient.  However, I recommend against `.dockerignore` for a beginners tutorial.  It's more complicated, and raises the potential of accidentally including a large dataset that the researchers don't want.  It's not uncommon for researchers workspace to contain many folders that are not relevant to their experiment.

```

> Copy the above Dockerfile to the same directory as your code, and modify it so it copies over all the files needed to run your experiment.  As an example, in the rest of this tutorial I will use the code in https://github.com/beaker/pytorch-example.

Reviewer: I like this example, but I wonder if the two of us should produce one for training an AllenNLP model -- both for my use in demos and for use here, as it'd mimic an actual workflow and probably be useful for pattern matching.

We don't have to do this now, but food for future thought.

> Now that you have a Dockerfile, you can build it with `docker build -t my-experiment .`.  This will take a while the first time because building the Docker image involves downloading and installing PyTorch, but the second time you build this will be cached and be much faster.  Once your image is built, you can try it out locally with `docker run -it --rm my-experiment python main.py`.  If your experiment requires a GPU (like the example does) it might fail when you run it locally.

Reviewer: ```suggestion
Now that you have a Dockerfile, you can build it with `docker build -t my-experiment .`.  This will take a while the first time because building the Docker image involves downloading and installing PyTorch, but the second time you build this will be cached and be much faster.  Once your image is built, you can try it out locally with `docker run -it --rm my-experiment python main.py`.  If your experiment requires a GPU (like the example does) it might fail when you run it locally.

The `my-example` name is the image tag. You can use any string you'd like -- we just use `my-experimtnet` here as an example.

```

> ## Creating a Beaker configuration for your experiment

Reviewer: ```suggestion
## Creating a Beaker Specification for your Experiment
```

> Since we've tried out our Docker image local, it's time to upload it to Beaker.  Since PyTorch is pretty big, it's best to do this from a location with fast upload speeds.

Reviewer: ```suggestion
Now that we've tried out running the Docker image locally, and we're relatively sure things work, it's time to upload it to Beaker.  Since PyTorch is pretty big, it's best to do this from a location with fast upload speeds.
```

```
beaker image create --name my-experiment my-experiment
```

Now that we have a Docker image we can begin to define our experiment.  We will do this by writing

> a YAML file `beaker-config.yaml` that tells Beaker information like what image to use and where to run the experiment.

Reviewer: ```suggestion
a YAML file called `beaker-config.yaml` that tells Beaker information like what image to use and where to run the experiment.
```

I might call it `beaker-spec.yaml`, as I *think* "spec" is the name we use for this in other Beaker documentation.

```

> version: v2-alpha

Reviewer: Just curious, when will v2 not be alpha? Do we have a migration path planned for that? @ckarenz @aaasen?

description: My Experiment
tasks:
  # We only have one step in our experiment, so there's only one entry in this list
  - name: training

>     image:

Reviewer: ```suggestion
   image:
```

      # You will want to replace `username` below with your Beaker username
      beaker: username/my-experiment
    arguments: [python, -u, main.py]
    result:
      # Beaker will capture anything that's written to this location and store it in the results
      # dataset.
      path: /output
    resources:
      gpuCount: 1
    context:
      cluster: ai2/on-prem-ai2-server
      priority: normal
```

For a more compilicated experiment you may need more configuration, perhaps to link together
multiple tasks or to mount in a dataset that your experiment needs.  For more details, please see
[the experiment spec](https://github.com/beaker/docs/blob/main/docs/concept/experiments.md#spec-format).

Now that we have our Docker image uploaded and our experiment spec written, we can run our job on
Beaker with a single command.

```
beaker experiment create beaker-conf.yaml
```

> That's it!  While that took a bit of setup, now moving our job to another cluster is simply a
> one line change, and it would be easy to script a large hyperparameter tuning.  Also, our results
> are persisted forever on Beaker.org, so if we ever need to check back on our experiment we can
> just look there.

Reviewer: ```suggestion
The command will output a beaker URL that you can visit to see details about your experiment.

That's it!  While that took a bit of setup, now moving our job to another cluster is simply a
one line change, and it would be easy to script a large hyperparameter tuning.  Also, our results
are persisted forever on Beaker.org, so if we ever need to check back on our experiment we can
just look there.
```
